File: backend/strategy/backtest.py
# backend/strategy/backtest.py
from strategy.condition import calculate_rsi, calculate_macd, calculate_ma

def run_backtest(
    data: list,          # 과거 OHLCV 데이터
    condition_type: str, # 매수 조건 (RSI, MACD, MA_CROSS)
    condition_value: float = 30,  # 조건값 (RSI의 경우 기준값)
    take_profit: float = 5.0,     # 익절 %
    stop_loss: float = 3.0,       # 손절 %
    order_amount: int = 100000,   # 1회 주문금액
) -> dict:
    """백테스트 실행"""

    if len(data) < 30:
        return {"error": "데이터가 부족해요 (최소 30일 필요)"}

    trades = []          # 매매 내역
    position = None      # 현재 보유 포지션
    cash = order_amount  # 시작 현금

    prices = [d["close"] for d in data]

    for i in range(20, len(data)):
        current = data[i]
        current_price = current["close"]
        current_date = current["date"]
        window = prices[:i+1]

        # 보유 중일 때 익절/손절 체크
        if position:
            profit_rate = ((current_price - position["buy_price"]) / position["buy_price"]) * 100

            # 익절
            if profit_rate >= take_profit:
                sell_amount = current_price * position["qty"]
                profit = sell_amount - position["buy_amount"]
                trades.append({
                    "type": "sell",
                    "reason": "익절",
                    "date": current_date,
                    "price": current_price,
                    "qty": position["qty"],
                    "profit": profit,
                    "profit_rate": profit_rate,
                    "hold_days": i - position["buy_idx"]
                })
                cash += sell_amount
                position = None
                continue

            # 손절
            if profit_rate <= -stop_loss:
                sell_amount = current_price * position["qty"]
                profit = sell_amount - position["buy_amount"]
                trades.append({
                    "type": "sell",
                    "reason": "손절",
                    "date": current_date,
                    "price": current_price,
                    "qty": position["qty"],
                    "profit": profit,
                    "profit_rate": profit_rate,
                    "hold_days": i - position["buy_idx"]
                })
                cash += sell_amount
                position = None
                continue

        # 포지션 없을 때 매수 조건 체크
        if not position:
            signal = False

            if condition_type == "RSI":
                rsi = calculate_rsi(window)
                if rsi and rsi <= condition_value:
                    signal = True

            elif condition_type == "MACD":
                macd, sig, hist = calculate_macd(window)
                if macd and sig and macd > sig:
                    signal = True

            elif condition_type == "MA_CROSS":
                ma5 = calculate_ma(window, 5)
                ma20 = calculate_ma(window, 20)
                if ma5 and ma20 and ma5 > ma20:
                    # 직전에 데드크로스였는지 확인
                    prev_window = prices[:i]
                    prev_ma5 = calculate_ma(prev_window, 5)
                    prev_ma20 = calculate_ma(prev_window, 20)
                    if prev_ma5 and prev_ma20 and prev_ma5 <= prev_ma20:
                        signal = True

            if signal:
                qty = max(1, int(order_amount / current_price))
                buy_amount = current_price * qty
                position = {
                    "buy_price": current_price,
                    "buy_date": current_date,
                    "buy_amount": buy_amount,
                    "buy_idx": i,
                    "qty": qty
                }
                cash -= buy_amount
                trades.append({
                    "type": "buy",
                    "date": current_date,
                    "price": current_price,
                    "qty": qty,
                    "amount": buy_amount
                })

    # 결과 분석
    sell_trades = [t for t in trades if t["type"] == "sell"]
    win_trades = [t for t in sell_trades if t["profit"] > 0]
    lose_trades = [t for t in sell_trades if t["profit"] <= 0]

    total_profit = sum(t["profit"] for t in sell_trades)
    total_profit_rate = (total_profit / order_amount) * 100 if sell_trades else 0
    win_rate = (len(win_trades) / len(sell_trades) * 100) if sell_trades else 0
    avg_hold_days = sum(t["hold_days"] for t in sell_trades) / len(sell_trades) if sell_trades else 0
    max_loss = min((t["profit_rate"] for t in sell_trades), default=0)

    return {
        "summary": {
            "total_trades": len(sell_trades),
            "win_trades": len(win_trades),
            "lose_trades": len(lose_trades),
            "win_rate": round(win_rate, 1),
            "total_profit": round(total_profit),
            "total_profit_rate": round(total_profit_rate, 2),
            "avg_hold_days": round(avg_hold_days, 1),
            "max_loss_rate": round(max_loss, 2),
        },
        "trades": trades
    }

File: frontend/src/pages/Backtest.js
// src/pages/Backtest.js
import React, { useState } from 'react';
import axios from 'axios';
import { LineChart, Line, XAxis, YAxis, CartesianGrid, Tooltip, ResponsiveContainer, ReferenceLine } from 'recharts';

const API = 'http://127.0.0.1:8080';

function Backtest() {
  const [ticker, setTicker] = useState('005930');
  const [conditionType, setConditionType] = useState('RSI');
  const [conditionValue, setConditionValue] = useState(30);
  const [takeProfit, setTakeProfit] = useState(5.0);
  const [stopLoss, setStopLoss] = useState(3.0);
  const [orderAmount, setOrderAmount] = useState(100000);
  const [result, setResult] = useState(null);
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState('');

  const runBacktest = async () => {
    setLoading(true);
    setError('');
    setResult(null);
    try {
      const res = await axios.get(`${API}/backtest/${ticker}`, {
        params: {
          condition_type: conditionType,
          condition_value: conditionValue,
          take_profit: takeProfit,
          stop_loss: stopLoss,
          order_amount: orderAmount
        }
      });
      setResult(res.data);
    } catch (e) {
      setError('백테스트 실행 실패');
    } finally {
      setLoading(false);
    }
  };

  // 누적 수익 차트 데이터
  const getChartData = () => {
    if (!result) return [];
    let cumProfit = 0;
    return result.result.trades
      .filter(t => t.type === 'sell')
      .map(t => {
        cumProfit += t.profit;
        return {
          date: t.date,
          profit: Math.round(cumProfit),
          rate: t.profit_rate
        };
      });
  };

  return (
    <div>
      <h2 style={{ fontSize: '20px', marginBottom: '24px' }}>🔬 백테스트</h2>

      {/* 설정 */}
      <div className="card">
        <h2>전략 설정</h2>
        <div style={{ display: 'flex', gap: '12px', flexWrap: 'wrap', marginBottom: '16px' }}>
          <div style={{ flex: 1, minWidth: '120px' }}>
            <label style={{ fontSize: '12px', color: '#8b949e' }}>종목코드</label>
            <input value={ticker} onChange={e => setTicker(e.target.value)} placeholder="005930" />
          </div>
          <div style={{ flex: 1, minWidth: '120px' }}>
            <label style={{ fontSize: '12px', color: '#8b949e' }}>매수 조건</label>
            <select
              value={conditionType}
              onChange={e => setConditionType(e.target.value)}
              style={{
                width: '100%', padding: '8px 12px',
                background: '#0d1117', border: '1px solid #30363d',
                borderRadius: '6px', color: '#e6edf3', fontSize: '14px'
              }}
            >
              <option value="RSI">RSI</option>
              <option value="MACD">MACD 골든크로스</option>
              <option value="MA_CROSS">이동평균 골든크로스</option>
            </select>
          </div>
          {conditionType === 'RSI' && (
            <div style={{ flex: 1, minWidth: '120px' }}>
              <label style={{ fontSize: '12px', color: '#8b949e' }}>RSI 기준값 (이하 매수)</label>
              <input type="number" value={conditionValue} onChange={e => setConditionValue(e.target.value)} />
            </div>
          )}
          <div style={{ flex: 1, minWidth: '120px' }}>
            <label style={{ fontSize: '12px', color: '#8b949e' }}>익절 (%)</label>
            <input type="number" value={takeProfit} onChange={e => setTakeProfit(e.target.value)} />
          </div>
          <div style={{ flex: 1, minWidth: '120px' }}>
            <label style={{ fontSize: '12px', color: '#8b949e' }}>손절 (%)</label>
            <input type="number" value={stopLoss} onChange={e => setStopLoss(e.target.value)} />
          </div>
          <div style={{ flex: 1, minWidth: '120px' }}>
            <label style={{ fontSize: '12px', color: '#8b949e' }}>1회 주문금액 (원)</label>
            <input type="number" value={orderAmount} onChange={e => setOrderAmount(e.target.value)} />
          </div>
        </div>
        <button className="btn btn-primary" onClick={runBacktest} disabled={loading}>
          {loading ? '⏳ 분석 중...' : '▶ 백테스트 실행'}
        </button>
        {error && <p style={{ color: '#f85149', marginTop: '8px' }}>{error}</p>}
      </div>

      {/* 결과 */}
      {result && (
        <>
          {/* 요약 */}
          <div className="card">
            <h2>📊 결과 요약 ({result.data_period} / {result.total_days}일)</h2>
            <div style={{ display: 'flex', gap: '24px', flexWrap: 'wrap' }}>
              {[
                { label: '총 수익률', value: `${result.result.summary.total_profit_rate}%`, color: result.result.summary.total_profit_rate >= 0 ? '#3fb950' : '#f85149' },
                { label: '총 손익', value: `${result.result.summary.total_profit.toLocaleString()}원`, color: result.result.summary.total_profit >= 0 ? '#3fb950' : '#f85149' },
                { label: '총 매매', value: `${result.result.summary.total_trades}회`, color: '#e6edf3' },
                { label: '승률', value: `${result.result.summary.win_rate}%`, color: '#58a6ff' },
                { label: '승/패', value: `${result.result.summary.win_trades}승 ${result.result.summary.lose_trades}패`, color: '#e6edf3' },
                { label: '평균 보유일', value: `${result.result.summary.avg_hold_days}일`, color: '#e6edf3' },
                { label: '최대 손실률', value: `${result.result.summary.max_loss_rate}%`, color: '#f85149' },
              ].map((item, idx) => (
                <div key={idx}>
                  <div style={{ fontSize: '12px', color: '#8b949e' }}>{item.label}</div>
                  <div style={{ fontSize: '20px', fontWeight: 'bold', color: item.color }}>{item.value}</div>
                </div>
              ))}
            </div>
          </div>

          {/* 누적 수익 차트 */}
          {getChartData().length > 0 && (
            <div className="card">
              <h2>📈 누적 손익 차트</h2>
              <ResponsiveContainer width="100%" height={200}>
                <LineChart data={getChartData()}>
                  <CartesianGrid strokeDasharray="3 3" stroke="#30363d" />
                  <XAxis dataKey="date" tick={{ fill: '#8b949e', fontSize: 11 }} />
                  <YAxis tick={{ fill: '#8b949e', fontSize: 11 }} />
                  <Tooltip
                    contentStyle={{ background: '#161b22', border: '1px solid #30363d' }}
                    formatter={(val) => [`${val.toLocaleString()}원`, '누적손익']}
                  />
                  <ReferenceLine y={0} stroke="#8b949e" />
                  <Line type="monotone" dataKey="profit" stroke="#58a6ff" dot={false} strokeWidth={2} />
                </LineChart>
              </ResponsiveContainer>
            </div>
          )}

          {/* 매매 내역 */}
          <div className="card">
            <h2>📋 매매 내역</h2>
            <table>
              <thead>
                <tr>
                  <th>날짜</th>
                  <th>구분</th>
                  <th>가격</th>
                  <th>수량</th>
                  <th>손익</th>
                  <th>수익률</th>
                  <th>사유</th>
                </tr>
              </thead>
              <tbody>
                {result.result.trades.map((t, idx) => (
                  <tr key={idx}>
                    <td style={{ color: '#8b949e', fontSize: '12px' }}>{t.date}</td>
                    <td style={{ color: t.type === 'buy' ? '#3fb950' : '#f85149' }}>
                      {t.type === 'buy' ? '매수' : '매도'}
                    </td>
                    <td>{t.price.toLocaleString()}원</td>
                    <td>{t.qty}주</td>
                    <td style={{ color: t.profit >= 0 ? '#3fb950' : '#f85149' }}>
                      {t.profit ? `${Math.round(t.profit).toLocaleString()}원` : '-'}
                    </td>
                    <td style={{ color: t.profit_rate >= 0 ? '#3fb950' : '#f85149' }}>
                      {t.profit_rate ? `${t.profit_rate.toFixed(2)}%` : '-'}
                    </td>
                    <td style={{ color: '#8b949e' }}>{t.reason || 'RSI 조건'}</td>
                  </tr>
                ))}
              </tbody>
            </table>
          </div>
        </>
      )}
    </div>
  );
}

export default Backtest;